shell/common/shell.cc

```diff
@@ -372,6 +371,71 @@ void Shell::NotifyLowMemoryWarning() const {
   // to purge them.
 }
 
+void Shell::RunEngine(RunConfiguration run_configuration) {
+  RunEngine(std::move(run_configuration), nullptr);
+}
+
+void Shell::RunEngine(RunConfiguration run_configuration,
+                      std::function<void(Engine::RunStatus)> result_callback) {
+  FML_DCHECK(is_setup_);
+  FML_DCHECK(task_runners_.GetPlatformTaskRunner()->RunsTasksOnCurrentThread());
+
+  if (!weak_engine_) {
+    result_callback(Engine::RunStatus::Failure);
+  }
+  fml::TaskRunner::RunNowOrPostTask(
+      task_runners_.GetUITaskRunner(),
+      fml::MakeCopyable([run_configuration = std::move(run_configuration),
+                         weak_engine = weak_engine_,
+                         result_callback =
+                             std::move(result_callback)]() mutable {
+        if (!weak_engine) {
+          if (result_callback) {
+            FML_LOG(ERROR)
+                << "Could not launch engine with configuration - no engine.";
+            result_callback(Engine::RunStatus::Failure);
+          }
+          return;
+        }
+        auto result = weak_engine->Run(std::move(run_configuration));
+        if (result == flutter::Engine::RunStatus::Failure) {
+          FML_LOG(ERROR) << "Could not launch engine with configuration.";
+        }
+        if (result_callback) {
+          result_callback(result);
+        }
+      }));
+}
+
+std::optional<int> Shell::GetUIIsolateLastError() const {
+  FML_DCHECK(is_setup_);
+  FML_DCHECK(task_runners_.GetPlatformTaskRunner()->RunsTasksOnCurrentThread());
+
+  if (!weak_engine_) {
+    return std::nullopt;
+  }
+  switch (weak_engine_->GetUIIsolateLastError()) {
+    case tonic::kCompilationErrorType:
+      return kCompilationErrorExitCode;
+    case tonic::kApiErrorType:
+      return kApiErrorExitCode;
+    case tonic::kUnknownErrorType:
+      return kErrorExitCode;
+    default:
+      return 0;
```

Comment:
`0` is actually `EXIT_SUCCESS`. So we will end up saying an unknown failure is no failure. This seems dodgy. Lets use `kErrorExitCode` as the default.

Reply:
I was under the impression the engine method returned 0 if ther ewas no error.  Is that not the case? 

The tester_main was defaulting this case to 0 as well AFAICT

Reply:
It seems to be `tonic::kNoError`. You can return `0` in that case instead of just defaulting to success.